### PR TITLE
Speedy: run() dont step()

### PR DESCRIPTION
**Pull request by nickchapman-da**
Run the Speedy machine with  `run()` instead of `step()`
- Remove: `SResultContinue`
- Add: `SResultFinalValue(_)`

We change the top level control of Speedy: from machine.step() to machine.run, with the control of stepping while the machine returns SResultContinue moved into speedy itself. (And so SResultContinue is removed in favour of SResultFinalValue.) The main advantage of this approach is that the tight while loop can be moved inside the exception handler, rather than having to wrap the handler every step.

changelog_begin
changelog_end

Benchmarked against master with `CollectAuthority`.  Speedup x1.09

this branch
```
Result "com.daml.lf.speedy.perf.CollectAuthority.bench":
  125.359 ±(99.9%) 1.410 ms/op [Average]
  (min, avg, max) = (121.084, 125.359, 131.553), stdev = 2.506
  CI (99.9%): [123.949, 126.769] (assumes normal distribution)
```

master
```
Result "com.daml.lf.speedy.perf.CollectAuthority.bench":
  136.720 ±(99.9%) 1.752 ms/op [Average]
  (min, avg, max) = (132.623, 136.720, 150.845), stdev = 3.114
  CI (99.9%): [134.968, 138.471] (assumes normal distribution)
```

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala [hunk 1]
```diff
@@ -432,7 +432,7 @@ object Repl {
             println(s"time: ${diff}ms")
             if (!errored) {
               val result = machine.ctrl match {
-                case Speedy.CtrlValue(sv) =>
+                case CtrlValue(sv) =>
```
Comment (hurryabit) [inline daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala:435]: I prefer to leave things like this in actual feature PRs unchanged. If you can't resist your temptation to change it, please put it in a separate PR which does nothing but these style corrections. Having two separate PRs makes reviewing easier.

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala [hunk 2]
```diff
@@ -48,9 +48,10 @@ final case class ScenarioRunner(
     var steps = 0
     while (!machine.isFinal) {
       //machine.print(steps)
-      machine.step match {
-        case SResultContinue =>
-          steps += 1
+      steps += 1 // this counts the number of external `Need` interactions
+      val res: SResult = machine.run()
+      res match {
+        case SResultFinalValue(_) =>
```
Comment (hurryabit) [inline daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala:54]: I don't see a consequence for this match branch.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala [hunk 3]
```diff
@@ -219,24 +219,17 @@ object Converter {
         seeding = InitialSeeding.NoSeed,
         Set.empty,
       )
-    @tailrec
-    def iter(): Either[String, (SValue, SValue)] = {
-      if (machine.isFinal) {
```
Comment (hurryabit) [inline daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala:224]: Why do we unconditionally run the machine now?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala [hunk 4]
```diff
@@ -18,7 +18,9 @@ sealed abstract class SResult extends Product with Serializable
 
 object SResult {
   final case class SResultError(err: SError) extends SResult
-  final case object SResultContinue extends SResult
+
+  /** The speedy machine has completed evaluation to reach a final value.  */
+  final case class SResultFinalValue(v: SValue) extends SResult
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:23]: Having `SResultFinalValue` carry the value as payload avoids having the caller dig around for it in the `Ctrl` of the machine, and knowing internal details of the machine's operation.

I think most callers dont really care about the final value because they are running the machine only for it's effect, or for other components, i.e. `Engine.scala` cares about `machine.ptx`.

There is one caller that makes use the payload:
 daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 5]
```diff
@@ -117,31 +117,33 @@ object Speedy {
...
-          res
-
-        case serr: SError =>
-          serr match {
-            case _: SErrorDamlException if tryHandleException =>
-              SResultContinue
-            case _ => SResultError(serr)
+    /** Run a machine until we get a result: either a final-value of a request for data, with a callback */
+    def run(): SResult = {
+      def again(): SResult = {
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:122]: `again()` is invoked in two places

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 6]
```diff
@@ -117,31 +117,33 @@ object Speedy {
...
+        // note: exception handler is outside while loop
+        try {
+          while (!isFinal) {
+            ctrl.execute(this) // make a single step
           }
-
-        case ex: RuntimeException =>
-          SResultError(SErrorCrash(s"exception: $ex"))
-
+          ctrl match {
```
[the remark this answers is not in the document]
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:128]: Because `SResultContinue` had different semantics.
Reusing an old name with changed semantics can be confusing.
Also I wanted there to be a payload. See above :)

File: daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala (hunk 1 above)
Comment (nickchapman-da) [reply daml-lf/repl/src/main/scala/com/digitalasset/daml/lf/repl/testing/Main.scala:435]: This was unintended.
An accidental hangover from splitting PRs. I'll revert this change back.

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala:54]: This is legal scala syntax. It means the same as having `()` on RHS of the `=>`
It's used all over the scala code base.
But I agree, I generally prefer when code is explicit with `()`, but I missed it here.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala (hunk 3 above)
Comment (nickchapman-da) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala:224]: I don't believe anything has changed.
It was being unconditionally run before. `step`ing until `isFinal`

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (hurryabit) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SResult.scala:23]: That makes sense. Thanks.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 7]
```diff
@@ -117,31 +117,33 @@ object Speedy {
...
+              SResultFinalValue(value) //stop
+            }
+            case _ =>
+              throw SErrorCrash(s"Unexpected ctrl on final machine $ctrl")
+          }
+        } catch {
+          case SpeedyHungry(res: SResult) => res //stop
+          case serr: SError =>
+            serr match {
+              case _: SErrorDamlException if tryHandleException => again() // not tail recursive
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:139]: Oh, I missed that call to `again`. We need to avoid recursion here. I don't want to blow through the stack at runtime.

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala (hunk 2 above)
Comment (hurryabit) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala:54]: Since that code has changed anyway, can we please add the `()` here.

File: daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala (hunk 3 above)
Comment (hurryabit) [reply daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Converter.scala:224]: That makes sense. Thanks.

File: daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala (hunk 2 above)
Comment (nickchapman-da) [reply daml-lf/scenario-interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/ScenarioRunner.scala:54]: ok

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 7 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala]: fixed

**Comment by nickchapman-da** [comment]
made requested changes

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 8]
```diff
@@ -117,31 +117,35 @@ object Speedy {
...
-      } catch {
-        case SpeedyHungry(res: SResult) =>
-          res
-
-        case serr: SError =>
-          serr match {
-            case _: SErrorDamlException if tryHandleException =>
-              SResultContinue
-            case _ => SResultError(serr)
+    /** Run a machine until we get a result: either a final-value of a request for data, with a callback */
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:120]:
```suggestion
    /** Run a machine until we get a result: either a final-value or a request for data, with a callback */
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 9]
```diff
@@ -117,31 +117,35 @@ object Speedy {
...
+    def run(): SResult = {
+      var result: SResult = null
+      while (result == null) {
+        // note: exception handler is outside while loop
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:124]: I think it would be useful to explain why there are two loops.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 8 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:120]: done in next PR

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 9 above)
Comment (nickchapman-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:124]: done in next PR

**Comment by hurryabit** [comment]
This was part of https://github.com/digital-asset/daml/issues/5764.
